add `item` bulk forbid/dump/melt tool (#896)
DFHack/scripts: docs/item.rst

wear-level, quality, ...), and perform the bulk operations forbid, dump, melt,
and their inverses. By default, the tool does not consider artifacts. Outputs
the number of items that matched the filters and were modified.

Usage
-----

``item [ count | [un]forbid | [un]dump | [un]hide | [un]melt ] <filter options>``

Action names should be self explanatory. All actions other than ``help`` and
``count`` have implicit filters associated with them. For instance, ``item
forbid --unreachable`` will neither report nor test reachability of items that
are already forbidden.

Examples
--------

``item forbid --unreachable``
    Forbid all items that cannot be reached by any of your citizens.

``item unforbid --inside Cavern1 --type wood``
    Unforbid/reclaim all logs inside the burrow named "Cavern1" (Hint: use 3D
    flood-fill to create a burrow covering an entire cavern layer).

``item melt -t weapon -m steel --max-quality 3``
    Designate all steel weapons whose quality is at most superior for melting.

``item hide -t boulder --scattered``
    Hide all scattered boulders, i.e. those that are not in stockpiles.

Options
-------

``-n, --dry-run``

>     Get an accurate count of the items that would be affected, including the
>     implicit filters of the selected main action.

Reviewer: ```suggestion
    Get a count of the items that would be modified by an operation, which will be the
    number returned by the ``count`` action minus the number of items with the desired
    property already set.
```

``-a, --include-artifacts``
    Include artifacts in the item list. Regardless of this setting, artifacts
    are never dumped or melted.

``-i, --inside <burrow>``
    Only include items inside the given burrow.

``-o, --outside <burrow>``
    Only include items outside the given burrow.

``-r, --reachable``
    Only include items reachable by one of your citizens.

``-u, --unreachable``
    Only include items not reachable by any of your citizens.

``-t, --type <string>``
    Filter by item type (e.g., BOULDER, CORPSE, ...). Also accepts lower case
    spelling (e.g. "corpse"). Use ``:lua @df.item_type`` to get the list of all
    item types.

``-m, --material <string>``
    Filter by material the item is made out of (e.g., "iron").

``-c, --mat-category <string>``
    Filter by material category of the material item is made out of (e.g.,

>     "metal"). Use ``:lua df.dfhack_material_category`` to get a list of all

Reviewer: ```suggestion
    "metal"). Use ``:lua @df.dfhack_material_category`` to get a list of all
```

    material categories.

``-d, --description <pattern>``
    Filter by item description (singular form without stack sizes). The
    ``pattern`` is a Lua pattern

>     (cf. https://www.lua.org/manual/5.4/manual.html#6.4.1), so "cave spider

Reviewer: DFHack uses Lua 5.3. I don't think it matters for patterns, but it would still be good to point people to the docs that match the version we're using

    Only include items whose wear/damage level is at most ``integer``. Useful
    values are 0 (pristine) to 3 (XX).

``-q, --min-quality <integer>``
    Only include items whose quality level is at least ``integer``. Useful
    values are 0 (ordinary) to 5 (masterwork).

``-Q, --max-quality <integer>``
    Only include items whose quality level is at most ``integer``. Useful
    values are 0 (ordinary) to 5 (masterwork).

``--stockpiled``
    Only include items that are in stockpiles. Does not include empty bins,
    barrels, and wheelbarrows assigned as storage and transport for stockpiles.

``--scattered``
    Opposite of ``--stockpiled``

``--forbidden``
    Only include forbidden items.

``--melting``
    Only include items designated for melting.

``--dumping``
    Only include items designated for dumping.

``--visible``
    Only include visible items (i.e., ignore hidden items).

``--hidden``
    Only include hidden items.

API
---

The item script can be called programmatically by other scripts, either via the
commandline interface with ``dfhack.run_script()`` or via the API functions
defined in :source-scripts:`item.lua`, available from the return value of

> ``reqscript('item')``:
> 
> * ``item.execute(action, conditions, options)``
> * ``item.executeWithPrinting(action, conditions, options)``

Reviewer: ```suggestion
``reqscript('item')``::

    local itemtools = reqscript('item')
    itemtools.execute(action, conditions, options)
    itemtools.executeWithPrinting(action, conditions, options)
```

Reviewer: I'm reluctant to call the variable that holds the `reqscript` return value "item" in the example since that is used in so many scripts as a local reference to an actual item.

Performs ``action`` (``forbid``, ``melt``, etc.) on all items satisfying
``conditions`` (a table containing functions from item to boolean). ``options``
is a table containing the boolean flags ``artifact``, ``dryrun``, and
``bytype``, which correspond to the (filter) options described above.

> The function ``item.execute`` performs no output, while the ``WithPrinting``

Reviewer: ```suggestion
The function ``execute`` performs no output, while the ``WithPrinting``
```

Reviewer: or prefix with `itemtools`, that would be ok too

variant performs the same output as the ``item`` tool.

The API provides a number of helper functions to aid in the construction of the
filter table. The first argument ``tab`` is always the table to which the filter
should be added.

> * ``item.condition_burrow(tab,burrow, outside)``

Reviewer: ```suggestion
* ``item.condition_burrow(tab, burrow, outside)``
```

* ``item.condition_type(tab, match)``
    If ``match`` is a string, this corresponds to ``--type <match>``. Also
    accepts numbers, matching against ``item:getType()``

* ``item.condition_reachable(tab)``
    Corresponds to ``--reachable``

* ``item.condition_unreachable(tab)``
    Corresponds to ``--unreachable``

* ``item.condition_description(tab, pattern)``
    Corresponds to ``--description <pattern>``

* ``item.condition_material(tab, match)``
    Corresponds to ``--material <match>``

* ``item.condition_matcat(tab, match)``
    Corresponds to ``--mat-category <match>``

* ``item.condition_wear(tab, lower, upper)``
    Selects items with wear level between ``lower`` and ``upper`` (Range 0-3, see above).

* ``item.condition_quality(tab, lower, upper)``
    Selects items with quality between ``lower`` and ``upper`` (Range 0-5, see above).

* ``item.condition_stockpiled(tab, invert)``
    Selects stockpiled items, or scattered items when ``invert=true``.

* ``item.condition_forbidden(tab)``
    Checks for ``item.flags.forbid``

* ``item.condition_melt(tab)``
    Checks for ``item.flags.melt``

* ``item.condition_dump(tab)``

>     Checks for ``item.flags.dump``

Reviewer: shouldn't these take a boolean so the inverse can be matched?

Reviewer: also, there seem to be a mix of `condition.foo()` + `condition.notfoo()` and `condition.foo(bool)`. I'd prefer to standardize on the latter, taking a boolean to indicate whether the relevant property should be true or false. that is foo(true) should match foo being true. I'd prefer to avoid invert=true semantics, like what is currently on the `stockpiled` condition.

Author: I agree, the API for condition creation still looks somewhat haphazard and that `condition.foo(...,bool)` is the way to go. However, I would still advocate for `negate=true` semantics. This way, the argument can be naturally elided when the condition is to be matched positively, which is probably the majority of cases. Otherwise, this would require `true` and `nil` to behave the same and `false` to stand for negation.

* ``item.condition_hidden(tab)``
    Checks for ``item.flags.hidden``

* ``item.condition_visible(tab)``
    Checks for ``not item.flags.hidden``

 API usage example::

>    local item = reqscript('item')
>    local cond = {}
> 
>    item.condition_type(cond, "BOULDER")
>    item.execute('unhide', cond, {}) -- reveal all boulders
> 
>    item.condition_stockpiled(cond, true)
>    item.execute('hide', cond, {})   -- hide all boulders not in stockpiles

Reviewer: ```suggestion
   local itemtools = reqscript('item')
   local cond = {}

   itemtools.condition_type(cond, "BOULDER")
   itemtools.execute('unhide', cond, {}) -- reveal all boulders

   itemtools.condition_stockpiled(cond, true)
   itemtools.execute('hide', cond, {})   -- hide all boulders not in stockpiles
```

item
====

.. dfhack-tool::
    :summary: Perform bulk operations on items based on various properties.
    :tags: fort productivity items

Filter items in you fort by various properties (e.g., item type, material,
wear-level, quality, ...), and perform the bulk operations forbid, dump, melt,
and their inverses. By default, the tool does not act on artifacts.

Usage
-----

> ``item [ help | count | [un]forbid | [un]dump | [un]melt ] <filter option>``

Author: Yes, this is basically why I stayed away from this for the moment. Really want this though.

Options
-------

``-h, --help``
    Print help.

``-r, --reachable``
    Only include items reachable by one of your citizens.

``-u, --unreachable``
    Only include items not reachable by one of your citizens.

``-t, --type <string>``
    Filter by item type (e.g., BOULDER, CORPSE, ...). Also accepts lower case
    spelling (e.g. "corpse")

``-m, --material <string>``
    Filter by material the item is made out of (e.g., "iron").

``-d, --description <string>``
    Filter by item description (singular form without stack sizes). Example:
    "cave spider silk web". Note: This does not work for animal products such as

>     wool, because their description always includes the stack size.

Author: I reformulated this to make it more clear.

item
====

.. dfhack-tool::
    :summary: Perform bulk operations on items based on various properties.
    :tags: fort productivity items

Filter items in you fort by various properties (e.g., item type, material,
wear-level, quality, ...), and perform the bulk operations forbid, dump, melt,
and their inverses. By default, the tool does not consider artifacts. Outputs
the number of items that matched the filters and were modified.

Usage
-----

``item [ help | count | [un]forbid | [un]dump | [un]melt ] <filter option>``

Action names should be self explanatory. All actions other than ``help`` and
``count`` have implicit filters associated with them. For instance, ``item
forbid --unreachable`` will neither report nor test reachability of items that
are already forbidden.

> 

Reviewer: needs "Examples" section here, before "Options". it would be a good idea to include a `--description` example for dealing with wool so the warning below has something to point to for how to do it "right"

Options
-------

``-h, --help``
    Print help.

> 

Reviewer: while it's usually a good idea to implement --help, help options are not usually documented since the canonical way to get tool help is `help toolname`

``-n, --dry-run``
    Use get an accurate count of the items that would be affected, including the
    implicit filters of the selected main action.

``-i, --inside <burrow>``
    Only include items inside the given burrow.

``-o, --outside <burrow>``
    Only include items outside the given burrow.

``-r, --reachable``
    Only include items reachable by one of your citizens.

``-u, --unreachable``

>     Only include items not reachable by one of your citizens.

Reviewer: ```suggestion
    Only include items not reachable by any of your citizens.
```

``-t, --type <string>``
    Filter by item type (e.g., BOULDER, CORPSE, ...). Also accepts lower case
    spelling (e.g. "corpse"). Use ``:lua @df.item_type`` to get the list of all
    item types.

``-m, --material <string>``
    Filter by material the item is made out of (e.g., "iron").

``-d, --description <string>``
    Filter by item description (singular form without stack sizes). Example:
    "cave spider silk web". Note: Due to a bug, this is of limited use for some
    animal products such as wool, because their description always includes the
    stack size.

``-a, --include-artifacts``
    Include artifacts in the item list. Regardless of this setting, artifacts
    are never dumped or melted

> ``--min-wear <integer>``

Reviewer: wear limits are actually frequently used and should probably be given short options. maybe `-w` and `-W`?

    Only include items whose wear/damage level is at least ``integer``. Useful
    values are 0 (pristine) to 3 (XX).

> ``--min-wear <integer>``

Reviewer: ```suggestion
``--max-wear <integer>``
```

    Only include items whose wear/damage level is at most ``integer``. Useful
    values are 0 (pristine) to 3 (XX).

> ``--min-quality <integer>``

Reviewer: quality options are frequently used as well and could use short forms, say `-q` and `-Q`

    Only include items whose quality level is at least ``integer``. Useful
    values are 0 (standard) to 5 (masterwork).

> ``--min-quality <integer>``

Reviewer: ```suggestion
``--max-quality <integer>``
```

    Only include items whose quality level is at most ``integer``. Useful

>     values are 0 (standard) to 5 (masterwork).

Reviewer: since you have a public API, it should be documented in an `API` section here

Reviewer: The script API still needs documenting; see https://docs.dfhack.org/en/stable/docs/tools/quickfort.html#api for an example

item
====

.. dfhack-tool::
    :summary: Perform bulk operations on items based on various properties.
    :tags: fort productivity items

Filter items in you fort by various properties (e.g., item type, material,
wear-level, quality, ...), and perform the bulk operations forbid, dump, melt,
and their inverses. By default, the tool does not consider artifacts. Outputs
the number of items that matched the filters and were modified.

Usage
-----

> ``item [ count | [un]forbid | [un]dump | [un]melt ] <filter options>``

Reviewer: there have been requests on Reddit for a tool to hide all boulders not in stockpiles. that could be worked into this tool pretty easily, I think. new subcommands [un]hide, new condition: inside/outside stockpile

Author: Indeed, adding `[un]hide` should be easy. That's just a flag, isn't it? 

Regarding stockpiles: The use-case you outlined requires a (fast) test whether an item is in *any* stockpile. I'm not sure whats the best way to go about this.

Reviewer: I think a scan of all stockpiles and a hashmap might be required. we have `dfhack.buildings.getStockpileContents(stockpile)` to get the list of stockpile contents, so iterating over all stockpiles and calling that would do it. It shouldn't be much worse than the existing iteration over all items.

Reviewer: also have to iterate over the contents of containers returned by `getStockpileContents`

Author: That, and iterate over the contents of containers (i.e. bags) contained in containers. Seems to be working reasonably fast now, even without jumping through extra hoops.

> ``item [ help | count | [un]forbid | [un]dump | [un]melt ] <filter option>``

Reviewer: ```suggestion
``item [ count | [un]forbid | [un]dump | [un]melt ] <filter options>``
```

Reviewer: supporting `help` as a command is fine, but the canonical way to get help is `help item`, so we don't want to promote alternatives.

Action names should be self explanatory. All actions other than ``help`` and
``count`` have implicit filters associated with them. For instance, ``item
forbid --unreachable`` will neither report nor test reachability of items that
are already forbidden.

Examples
--------

``item forbid --unreachable``
    Forbid all items that cannot be reached by any of your citizens.

``item unforbid --inside Cavern1 --type wood``
    Unforbid/reclaim all logs inside the burrow named "Cavern1" (Hint: use 3D

>     flood-fill to create a burrow covering an entire cavern layer)

Reviewer: ```suggestion
    flood-fill to create a burrow covering an entire cavern layer).
```

> ``item melt -t weapon -m steel --max-quality 3``

Reviewer: I suggest supporting words for quality levels (in addition to numbers) and using the words in the example here. People aren't going to remember the numeric mappings.

Author: I'm not even sure myself which words I should be accepting here. There isn't even perfect agreement between the wiki and DFHack: "Superior" vs "Superior quality". Some of the levels would have dashes (finely-crafted). Also, is it "masterful" or "masterwork". I think numbers are easier and easier to get right.

>     Designate all steel weapons whose quality is less than "exceptional" for

Reviewer: if the parameter is named "max', the wording here should be "at most" not "is less than"

    melting.

Options
-------

``-n, --dry-run``

>     Use get an accurate count of the items that would be affected, including the

Reviewer: ```suggestion
    Get an accurate count of the items that would be affected, including the
```

>     implicit filters of the selected main action.

Reviewer: the implicit filters need documenting : )

Author: There is already a paragraph right under "Usage". Are you saying that's not enough?

Reviewer: I'm not sure I'd call not reporting items that are already forbidden an "implicit filter". If `item forbid --unreachable` included the count of items that were already forbidden, I'd just call that a bug : P

What about items that are owned, in a job, rotten, on fire, part of a building, etc.? Many other tools implicitly filter these out. I thought this was more what you meant.

Reviewer: after reading the actual code, I see what you meant by "implicit filter" I suggested some rewording to make it (I think) clearer

``-i, --inside <burrow>``
    Only include items inside the given burrow.

``-o, --outside <burrow>``
    Only include items outside the given burrow.

``-r, --reachable``
    Only include items reachable by one of your citizens.

``-u, --unreachable``
    Only include items not reachable by one of your citizens.

``-t, --type <string>``
    Filter by item type (e.g., BOULDER, CORPSE, ...). Also accepts lower case
    spelling (e.g. "corpse"). Use ``:lua @df.item_type`` to get the list of all
    item types.

``-m, --material <string>``
    Filter by material the item is made out of (e.g., "iron").

``-c, --mat-category <string>``
    Filter by material category of the material item is made out of (e.g.,

>     "metal"). Use ``:lua df.dfhack_material_category`` to get have a list of all

Reviewer: ```suggestion
    "metal"). Use ``:lua df.dfhack_material_category`` to get a list of all
```

    material categories.

``-d, --description <string>``
    Filter by item description (singular form without stack sizes). Example:
    "cave spider silk web". Note: Due to a bug, this is of limited use for some
    animal products such as wool, because their description always includes the

>     stack size.

Reviewer: maybe give an example of what the wool description is and how to deal with it

Author: I don't think there's a good way to deal with "wool", at least not without changing the meaning of the description filter. The description provided by DF always involves the stack size. I guess one could work around this issue by removing any potential stack size at the end of the string provided by DF or matching the argument against parts of the description provided by DF.

Author: I ended up resolving this by (a) removing any trailing stack size that gets reported by DF and (b) using pattern/regex matching rather than string equality for the ``--description`` filter.

``-a, --include-artifacts``
    Include artifacts in the item list. Regardless of this setting, artifacts

>     are never dumped or melted

Reviewer: ```suggestion
    are never dumped or melted.
```

``-w, --min-wear <integer>``
    Only include items whose wear/damage level is at least ``integer``. Useful
    values are 0 (pristine) to 3 (XX).

``-W, --max-wear <integer>``
    Only include items whose wear/damage level is at most ``integer``. Useful
    values are 0 (pristine) to 3 (XX).

``-q, --min-quality <integer>``
    Only include items whose quality level is at least ``integer``. Useful

>     values are 0 (standard) to 5 (masterwork).

Reviewer: ```suggestion
    values are 0 (ordinary) to 5 (masterwork).
```

``-Q, --max-quality <integer>``
    Only include items whose quality level is at most ``integer``. Useful

>     values are 0 (standard) to 5 (masterwork).

Reviewer: ```suggestion
    values are 0 (ordinary) to 5 (masterwork).
```
